File: Server/src/main/java/com/github/dr/rwserver/net/netconnectprotocol/AbstractGameVersion.java
package com.github.dr.rwserver.net.netconnectprotocol;

import com.github.dr.rwserver.data.Player;
import com.github.dr.rwserver.data.global.Data;
import com.github.dr.rwserver.data.global.NetStaticData;
import com.github.dr.rwserver.io.GameInputStream;
import com.github.dr.rwserver.io.GameOutputStream;
import com.github.dr.rwserver.io.Packet;
import com.github.dr.rwserver.net.GroupNet;
import com.github.dr.rwserver.net.core.AbstractNetConnect;
import com.github.dr.rwserver.net.game.ConnectionAgreement;
import com.github.dr.rwserver.util.Time;
import com.github.dr.rwserver.util.log.Log;
import com.github.dr.rwserver.util.zip.gzip.GzipEncoder;
import okhttp3.internal.cache2.Relay;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;

/**
 * 作为{@link AbstractNetConnect} 和 协议实现的中间人
 * 目的是为了多协议支持
 * 共用协议放在本处
 * @author Dr
 */
public abstract class AbstractGameVersion implements AbstractNetConnect {
    /** 错误次数 */
    protected int errorTry = 0;
    /** 是否停留在输入界面 */
    protected boolean isPasswd = false;
    /** 最后一次接受到包的时间 */
    protected long lastReceivedTime = Time.concurrentMillis();
    /** 玩家是否死亡 */
    protected volatile boolean isDis = false;
    /** 是否已经重试过 */
    protected volatile boolean isTry = false;
    /** 玩家 */
    protected Player player = null;

    protected ConnectionAgreement connectionAgreement;

    @Override
    public String getIp() {
        return connectionAgreement.ip;
    }

    @Override
    public int getPort() {
        return connectionAgreement.localPort;
    }

    @Override
    public String getName() {
        return "";
    }

    @Override
    public void setCache(Packet packet) {
    }

    @Override
    public void setTry() {
        errorTry++;
    }

    @Override
    public int getTry() {
        return errorTry;
    }

    @Override
    public void setTryBoolean(boolean tryBoolean) {
        this.isTry = tryBoolean;
    }

    @Override
    public boolean getTryBoolean() {
        return isTry;
    }

    @Override
    public boolean getIsPasswd() {
        return isPasswd;
    }

    @Override
    public void setLastReceivedTime() {
        this.isTry = false;
        this.lastReceivedTime = Time.concurrentMillis();
    }

    @Override
    public long getLastReceivedTime() {
        return lastReceivedTime;
    }

    @Override
    public void setConnectionAgreement(ConnectionAgreement connectionAgreement) {
        this.connectionAgreement = connectionAgreement;
    }

    @Override
    public String getConnectionAgreement() {
        return connectionAgreement.useAgreement;
    }

    @Override
    public String getVersion() {
        return "1.14";
    }

    @Override
    public void sendSystemMessage(@NotNull String msg) {
        try {
            sendPacket(NetStaticData.protocolData.abstractNetPacket.getSystemMessagePacket(msg));
        } catch (IOException e) {
            Log.error("[Player] Send System Chat Error",e);
        }
    }

    @Override
    public void sendChatMessage(@NotNull String msg, String sendBy, int team) {
        try {
            sendPacket(NetStaticData.protocolData.abstractNetPacket.getChatMessagePacket(msg,sendBy,team));
        } catch (IOException e) {
            Log.error("[Player] Send Player Chat Error",e);
        }
    }

    @Override
    public void sendServerInfo(boolean utilData) throws IOException {
    }

    @Override
    public void sendSurrender() {
    }

    @Override
    public void sendKick(@NotNull String reason) throws IOException {
    }

    @Override
    public void ping() {
        try {
            sendPacket(NetStaticData.protocolData.abstractNetPacket.getPingPacket(player));
        } catch (IOException e) {
            errorTry++;
        }
    }

    @Override
    public byte[] getGameSaveData(@NotNull Packet packet) throws IOException {
        return null;
    }

    @Override
    public void receiveChat(@NotNull Packet p) throws IOException {
    }

    @Override
    public void receiveCommand(@NotNull Packet p) throws IOException {
    }

    @Override
    public void sendStartGame() throws IOException {
    }

    @Override
    public void sendTeamData(@NotNull GzipEncoder gzip) {
    }

    @Override
    public boolean getPlayerInfo(@NotNull Packet p) throws IOException {
        return false;
    }

    @Override
    public void registerConnection(@NotNull Packet p) throws IOException {
    }

    @Override
    public void sendErrorPasswd() throws IOException {
    }

    protected void close(final GroupNet groupNet) {
        try {
            connectionAgreement.close(groupNet);
        } catch (Exception e) {
            Log.error("Close Connect",e);
        }
    }

    @Override
    public void getGameSave() {
    }

    @Override
    public void sendGameSave(@NotNull Packet packet) {
        sendPacket(packet);
    }

    @Override
    public void sendPacket(@NotNull Packet packet) {
        try {
            connectionAgreement.send(packet);
        } catch (Exception e) {
            Log.error("[UDP] SendError - 本消息单独出现无妨 连续多次出现请debug",e);
            disconnect();
        }
    }
}

File: Server/src/main/java/com/github/dr/rwserver/net/netconnectprotocol/AbstractGameVersion.kt
package com.github.dr.rwserver.net.netconnectprotocol

import com.github.dr.rwserver.data.Player
import com.github.dr.rwserver.data.global.NetStaticData
import com.github.dr.rwserver.io.GameOutputStream
import com.github.dr.rwserver.io.Packet
import com.github.dr.rwserver.net.GroupNet
import com.github.dr.rwserver.net.core.server.AbstractNetConnect
import com.github.dr.rwserver.net.game.ConnectSrver
import com.github.dr.rwserver.net.game.ConnectionAgreement
import com.github.dr.rwserver.util.PacketType
import com.github.dr.rwserver.util.Time.concurrentMillis
import com.github.dr.rwserver.util.log.Log
import com.github.dr.rwserver.util.zip.gzip.GzipEncoder
import java.io.IOException

/**
 * 作为[AbstractNetConnect] 和 协议实现的中间人
 * 目的是为了多协议支持
 * 共用协议放在本处
 * @author Dr
 */
 abstract class AbstractGameVersion(@JvmField protected val connectionAgreement: ConnectionAgreement): AbstractNetConnect {
    /** 错误次数  */
    override var `try` = 0
        protected set

    /** 是否停留在输入界面  */
    override var inputPassword = false
        protected set

    /** 最后一次接受到包的时间  */
    override var lastReceivedTime = concurrentMillis()
        protected set

    override val isConnectServer: Boolean = false
    override var connectServer: ConnectSrver? = null

    /** 玩家是否死亡  */
    @JvmField
    @Volatile
    protected var isDis = false

    /** 是否已经重试过  */
    @Volatile
    override var tryBoolean = false

    /** 玩家  */
    override lateinit var player: Player


    override val ip: String
        get() = connectionAgreement.ip

    override val port: Int
        get() = connectionAgreement.localPort

    override val name: String
        get() = ""

    override fun setCache(packet: Packet) {}
    override fun setCacheA(packet: Packet) {}
    override fun setTry() {
        `try`++
    }

    override fun setLastReceivedTime() {
        tryBoolean = false
        lastReceivedTime = concurrentMillis()
    }

    override fun getConnectionAgreement(): String {
        return connectionAgreement.useAgreement
    }

    override val version: String
        get() = "1.14"

    override fun sendSystemMessage(msg: String) {
        try {
            sendPacket(NetStaticData.protocolData.abstractNetPacket.getSystemMessagePacket(msg))
        } catch (e: IOException) {
            Log.error("[Player] Send System Chat Error", e)
        }
    }

    override fun sendChatMessage(msg: String, sendBy: String, team: Int) {
        try {
            sendPacket(NetStaticData.protocolData.abstractNetPacket.getChatMessagePacket(msg, sendBy, team))
        } catch (e: IOException) {
            Log.error("[Player] Send Player Chat Error", e)
        }
    }

    @Throws(IOException::class)
    override fun sendServerInfo(utilData: Boolean) {
    }

    override fun sendSurrender() {}
    @Throws(IOException::class)
    override fun sendKick(reason: String) {
        val o = GameOutputStream()
        o.writeString(reason)
        sendPacket(o.createPacket(PacketType.PACKET_KICK))
        disconnect()
    }

    override fun ping() {
        try {
            sendPacket(NetStaticData.protocolData.abstractNetPacket.getPingPacket(player))
        } catch (e: IOException) {
            `try`++
        }
    }

    @Throws(IOException::class)
    override fun getGameSaveData(packet: Packet): ByteArray = ByteArray(0)

    @Throws(IOException::class)
    override fun receiveChat(p: Packet) {}

    @Throws(IOException::class)
    override fun receiveCommand(p: Packet) {}

    @Throws(IOException::class)
    override fun sendStartGame() {}

    override fun sendTeamData(gzip: GzipEncoder) {}

    override fun sendRelayServerType(msg: String) {}

    override fun sendRelayServerTypeReply(packet: Packet) {}

    @Throws(IOException::class)
    override fun getPlayerInfo(p: Packet): Boolean {
        return false
    }

    @Throws(IOException::class)
    override fun registerConnection(p: Packet) {
    }

    @Throws(IOException::class)
    override fun sendErrorPasswd() {
    }

    override fun getGameSave() {}

    override fun sendGameSave(packet: Packet) {
        sendPacket(packet)
    }

    override fun sendPacket(packet: Packet) {
        try {
            connectionAgreement.send(packet)
        } catch (e: Exception) {
            Log.error("[UDP] SendError - 本消息单独出现无妨 连续多次出现请debug", e)
            disconnect()
        }
    }

    protected fun close(groupNet: GroupNet?) {
        try {
            connectionAgreement.close(groupNet)
        } catch (e: Exception) {
            Log.error("Close Connect", e)
        }
    }
}